### PR TITLE
DrivAerML: 4x case revisits per epoch (max-train-batches=1576)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Human directive #3020 mentions "revisit cases multiple times per epoch (Transolver-3 style, ~400k points/case/epoch)." Currently with 394 cases and 50k points, each case is seen once per epoch (394 batches). With max-train-batches=1576 (4x), each case is revisited ~4 times per epoch with DIFFERENT random 50k-point subsamples each time. This is data augmentation via resampling — the model sees different spatial subsets of each car geometry each epoch. Eren #2910 tested 2x (788 batches). This tests 4x and 8x — more aggressive coverage, potentially finding deeper loss basins through better gradient coverage over the full surface.

## Current Baseline (DM)
- val_primary/surface_rel_l2_pct = **3.997%** (AdamW lr=5e-4, T_max=30, no-EMA, 4L/512d, NO gc, NO WD)
- External target: 3.71%

## Runs

**Run 1 — 4x revisits (max-train-batches=1576)**
```bash
SENPAI_MAX_EPOCHS=9999 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 1576 --max-eval-batches 200 --wandb_group dm_multi_revisit
```

**Run 2 — 8x revisits (max-train-batches=3152, Transolver-3 territory)**
```bash
SENPAI_MAX_EPOCHS=9999 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 3152 --max-eval-batches 200 --wandb_group dm_multi_revisit
```

**Run 3 — 1x control with full eval (paper-facing, no --max-eval-batches)**
```bash
SENPAI_MAX_EPOCHS=9999 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --wandb_group dm_multi_revisit
```

## Expected outcome
If 4x or 8x revisits reduce surface_rel_l2_pct below 3.997%, this becomes the new training recipe. Also report approximate epochs reached for each run to understand the throughput trade-off.

## Notes
- Do NOT add --grad-clip or --weight-decay — champion config has neither
- The key mechanism is random subsampling: each revisit of a case samples a DIFFERENT 50k-point subset, giving different gradient signal
- If 8x causes OOM, reduce to 6x (max-train-batches=2364)

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT bulma:

## Results

### Baseline
- **val_primary/surface_rel_l2_pct = 3.997%** (PR #2898, AdamW lr=5e-4, T_max=30, no-EMA, 4L/512d, 467 epochs with --max-eval-batches 200)
- External target: 3.71%

### Run 1 — 4x revisits (max-train-batches=1576)

| Metric | Value |
|--------|-------|
| **val_primary/surface_rel_l2_pct** | **9.382%** (epoch 36) |
| test_primary/surface_rel_l2_pct | 10.226% (best-checkpoint) |
| Total epochs | 160 (360-min budget) |
| Diverged at | ~epoch 60 |
| Peak memory | ~6 GB VRAM |
| W&B run | a5oa7wpt |

```bash
CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 1576 --max-eval-batches 200 --wandb-group dm_multi_revisit --agent bulma --wandb-name "bulma/dm-4x-revisit"
```

**Convergence:** Fast initial descent (45.3% → 9.38% in 36 epochs). Top-5 trough values: ep36=9.382%, ep45=9.384%, ep39=9.419%, ep35=10.197%, ep18=10.262%. Model found good basins at cosine troughs before divergence.

**Divergence:** Grad norms exploded ~epoch 60 (from ~1 to 432K). Model never recovered, oscillating 30-65% for remaining 100 epochs.

### Run 2 — 8x revisits (max-train-batches=3152)

| Metric | Value |
|--------|-------|
| **val_primary/surface_rel_l2_pct** | **15.211%** (epoch 10) |
| test_primary/surface_rel_l2_pct | 15.905% (best-checkpoint) |
| Total epochs | ~88 (360-min budget) |
| Diverged at | ~epoch 12 |
| Peak memory | ~6 GB VRAM |
| W&B run | xsmxmlg7 |

```bash
CUDA_VISIBLE_DEVICES=1 SENPAI_MAX_EPOCHS=9999 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 3152 --max-eval-batches 200 --wandb-group dm_multi_revisit --agent bulma --wandb-name "bulma/dm-8x-revisit"
```

**Divergence:** Grad norms exploded by epoch 12 (from ~4 to 805, then 13K+). Reached NaN by epoch 79. 8x is too aggressive without grad clipping — 3152 gradient steps per cosine cycle (vs 394 for 1x) accumulates catastrophic instability.

### Run 3 — 1x control with full eval (no --max-eval-batches)

| Metric | Value |
|--------|-------|
| **val_primary/surface_rel_l2_pct** | **13.950%** (epoch 43) |
| test_primary/surface_rel_l2_pct | 15.709% (final-model, not best-checkpoint) |
| Total epochs | 46 (360-min budget) |
| Diverged | Never — stable throughout |
| Peak memory | ~6 GB VRAM |
| W&B run | itlteubz |

```bash
CUDA_VISIBLE_DEVICES=2 SENPAI_MAX_EPOCHS=9999 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --wandb-group dm_multi_revisit --agent bulma --wandb-name "bulma/dm-1x-control-fulleval"
```

**Convergence:** Very stable — grad norms 0.9-1.2 throughout all 46 epochs. Trough values consistently improved: 25.2% → 22.9% → 18.1% → 16.5% → 15.5% → 14.2% → 13.95%. Still descending at timeout. However, removing --max-eval-batches makes each epoch ~10x slower (~8 min/epoch vs ~0.77 min with --max-eval-batches 200), limiting total epochs to 46 vs 467 for the baseline.

### Summary

| Run | Best val % | Best epoch | Test % | Diverged? | vs Baseline (3.997%) |
|-----|-----------|-----------|--------|-----------|---------------------|
| 4x (1576) | 9.382 | 36 | 10.226 | Yes, ep~60 | +135% worse |
| 8x (3152) | 15.211 | 10 | 15.905 | Yes, ep~12 | +281% worse |
| 1x control (394, full eval) | 13.950 | 43 | 15.709 | No | +249% worse |

**None of the runs beat the baseline of 3.997%.**

## What happened

The multi-revisit hypothesis **did not work**. Two fundamental problems:

1. **Instability without grad clipping:** Both 4x and 8x revisits diverged because the additional gradient steps per epoch (1576/3152 vs 394) accumulate instability at cosine LR peaks. Without --grad-clip, the optimizer cannot survive the compounded gradient noise from revisiting the same cases with different subsamples at high learning rates. The 4x run survived 36 epochs before divergence; the 8x diverged within 12 epochs.

2. **Eval cost of full eval:** The 1x control (Run 3) without --max-eval-batches was epoch-starved — only 46 epochs vs 467 for the baseline in the same 360-min budget. The full eval over all cases is ~10x more expensive than the 200-batch eval used in the baseline config. This shows that --max-eval-batches 200 is essential for training efficiency.

**Key observation:** The 4x run's trough values (9.38%) were actually competitive at their epoch count — the baseline was at ~12-13% at epoch 36 as well. The problem is that the 4x config can't survive long enough to reach the deeper basins found at epoch 200-467. Multi-revisit trades stability for early convergence speed, but the stability loss is fatal.

## Suggested follow-ups

1. **4x revisits + --grad-clip 1.0:** The 4x run showed promising early convergence before divergence. Adding gradient clipping could stabilize it enough to reach the deeper basins. The champion config already has no --grad-clip, but multi-revisit may specifically need it.

2. **4x revisits + lower LR (e.g. 2.5e-4):** Halving the LR with 4x batches would give roughly the same effective gradient signal per epoch while reducing peak instability at cosine peaks.

3. **Progressive revisit schedule:** Start with 1x for the first 100 epochs (stable descent), then increase to 2x or 4x for final 100 epochs (data augmentation in the convergence phase when grad norms are naturally lower).

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3051 because ADVISOR: Closing — multi-revisit training is a dead end on DrivAerML in current form. 4x revisits best val 9.382% (135% above baseline), 8x best 15.211% (281% above), 1x full-eval control 13.950% (249% above). All diverged or were epoch-starved. The early convergence speed of 4x was promising but the config cannot survive to the epoch 200-467 range where baseline achieves 3.997%. Good systematic sweep and useful finding that --max-eval-batches 200 is well-calibrated, bulma..
